presto-main/src/main/java/io/prestosql/execution/StageStateMachine.java

```diff
@@ -107,7 +106,7 @@ public StageStateMachine(
         this.tables = ImmutableMap.copyOf(requireNonNull(tables, "tables is null"));
         this.scheduledStats = requireNonNull(schedulerStats, "schedulerStats is null");
 
-        stageState = new StateMachine<>("stage " + stageId, executor, PLANNED, TERMINAL_STAGE_STATES);
+        stageState = new StateMachine<>("stage " + stageId, executor, PLANNED, StageState::isDone);
```

Comment:
This is not related to QueryStateMachine changes.
In particular, the QueryStateMachine commit fixes a bug, but this change is just a refactor. They shouldn't be together in one commit.

Extract a preparatory commit with StateMachine change (accepting Predicate) and apply it here.
Then the rest.